src/plots/cartesian/select.js

```diff
@@ -349,11 +323,246 @@ function prepSelect(e, startX, startY, dragOptions, mode) {
     };
 }
 
+// Missing features
+// ----------------
+// TODO handle clearing selection when no point is clicked (based on hoverData)
+// TODO remove console.log statements
+function selectOnClick(gd, numClicks, evt, xAxes, yAxes, outlines) {
+    var hoverData = gd._hoverdata;
+    var retainSelection = shouldRetainSelection(evt);
+    var searchTraces;
+    var searchInfo;
+    var trace;
+    var clearEntireSelection;
+    var clickedPts;
+    var clickedPt;
+    var shouldSelect;
+    var traceSelection;
+    var allSelectionItems;
+    var eventData;
+    var i;
+    var j;
+
+    if(numClicks === 1 && isHoverDataSet(hoverData)) {
+        allSelectionItems = [];
+
+        searchTraces = determineSearchTraces(gd, xAxes, yAxes);
+        clearEntireSelection = entireSelectionToBeCleared(searchTraces, hoverData);
+
+        for(i = 0; i < searchTraces.length; i++) {
+            searchInfo = searchTraces[i];
+            trace = searchInfo.cd[0].trace;
+
+            // Clear old selection if needed
+            if(!retainSelection || clearEntireSelection) {
+                searchInfo._module.toggleSelected(searchInfo, false);
+                if(outlines) outlines.remove();
+
+                if(clearEntireSelection) continue;
+            }
+
+            // Determine clicked points,
+            // call selection modification functions of the trace's module
+            // and collect the resulting set of selected points
+            clickedPts = clickedPtsFor(searchInfo, hoverData);
```

Comment:
Does the `clickedPts`  array ever have more than one item?

Reply:
At the moment not. I changed it back when I was about to integrate the `box` trace type. See https://github.com/plotly/plotly.js/pull/2824/commits/231fa6d1933597bebaa271b5434f9f0cf56df8ca.

I figured it'd be a good idea to be ready for `hoverdata` one day containing more than one valid data point. E.g. when multiple data points are at the same location that was clicked.

Is allocating an array costly in terms of performance? Happy to change it back to non-array style if that is the case! Or, if we want to keep it, I could add a comment!

Reply:
> Is allocating an array costly in terms of performance?

It _can_ be costly, but I'll have to test on some real-world data to know for sure.

Reply:
For now I postpone moving back to a non-array return type until there's proof that it hurts performance considerably.